🤖 Add authors and contributors to Practice Exercises

ErikSchierboom opened the pull request:
With v3, we've introduced the idea of exercise authors and contributors. This information is stored in the exercise's `.meta/config.json` file (see [the docs](https://github.com/exercism/docs/blob/main/anatomy/tracks/practice-exercises.md#file-metaconfigjson)).

Concept Exercises, which are all new, have all defined their authors and contributors. Practice Exercises though are missing this information. In this PR, we attempt to populate the authors and contributors of Practice Exercises based on their commit history.

## Algorithm

We start out by looking at the current Practice Exercises. For each Practice Exercise, we'll look at the commit history of its files to see which commits did something with that Practice Exercise. How renames are handled is discussed in the "Renames" section later in this document. 

Each commit that we can link to a Practice Exercise will be used to determine authorship/contributorship. Here is how we assign authorship/contributorship:

### Authors

1. Find the Git commit author of the oldest commit linked to that Practice Exercise. As there can only be one oldest commit, we only have one author
2. Find the GitHub username for the Git commit author of that commit
3. Add the GitHub username of the Git commit author to the `authors` key in the `.meta/config.json` file

### Contributors

1. Find the Git commit authors of all but the oldest commit linked to that Practice Exercise. If there is only one commit, there won't be any contributors.
1. Exclude the Git commit author of the oldest commit from the list of Git commit authors (an author is _not_ also a contributor) 
2. Find the GitHub usernames for the Git commit authors of those commits
3. Add the GitHub usernames of the Git commit authors to the `contributor` key in the `.meta/config.json` file

We're using the GitHub GraphQL API to find the username of a commit author. In some cases though, a username cannot be found for a commit (e.g. due to the user account no longer existing), in which case we'll skip the commit. 

## Renames

There is a select number of Practice Exercises that might have been renamed at some point. You can ask Git to "follow" a file over its renames using `git log --follow <file>`, which will also return commits made before renames. Unfortunately, Git does not store renames, it just stores the contents of the renamed files and tries to guess if a file was renamed by looking at its contents. This _can_ (and will) lead to false positives, where Git will think a file has been renamed whereas it hasn't. As we don't want to have incorrect authors/contributors for exercises, we're ignoring renames. The only exception to this are known exercise renames:

- `bracket-push` was renamed to `matching-brackets`
- `retree` was renamed to `satellite`
- `resistor-colors` was renamed to `resistor-color-duo`
- `kindergarden-garden` was renamed to `kindergarten-garden`

If you know of a rename of a Practice Exercise, please check to see if its authors and contributors are correctly set.

## Exclusions

There are some commits that we skip over, which thus don't influence the authors/contributors list:

- Commits authored by `dependabot[bot]`, `dependabot-preview[bot]` or `github-actions[bot]`
- Bulk update PRs made by `ErikSchierboom` or `kytrinx` to update the track

## Maintainer TODO list

These are the things you, the maintainers, should do:

- Check the authors and contributors, adding/removing users where the data is missing or incorrect
- Double-check any renamed Practice Exercises
- Check for Practice Exercises with missing authors, which is most likely due to the author not being on GitHub anymore 
- Add additional authors if you know that a Practice Exercise has actually been created by more than one author 

If you find something needs updating, just push a new commit to this PR.

## Tracking

https://github.com/exercism/v3-launch/issues/24

SaschaMann commented on exercises/practice/accumulate/.meta/config.json:
```diff
@@ -1,10 +1,35 @@
 {
   "blurb": "Implement the `accumulate` operation, which, given a collection and an operation to perform on each element of the collection, returns a new collection containing the result of applying that operation to each element of the input collection.",
-  "authors": [],
+  "authors": [
+    {
+      "github_username": "SaschaMann"
+    }
+  ],
+  "contributors": [
+    {
+      "github_username": "kytrinyx"
+    },
+    {
+      "github_username": "fsouza"
+    },
+    {
+      "github_username": "Teo-ShaoWei"
+    },
+    {
+      "github_username": "logankilpatrick"
+    },
+    {
+      "github_username": "guilhermebodin"
+    }
+  ],
```
This is completely wrong, see https://github.com/exercism/julia/commits/v2-final-version/exercises/accumulate

Maybe because the exercise has been deprecated?

SaschaMann commented on exercises/practice/transpose/.meta/config.json:
```diff
@@ -1,10 +1,32 @@
 {
   "blurb": "Take input text and output it transposed.",
-  "authors": [],
+  "authors": [
+    {
+      "github_username": "andrej-makarov-skrt"
+    }
+  ],
+  "contributors": [
+    {
+      "github_username": "SaschaMann"
+    },
+    {
+      "github_username": "kytrinyx"
+    },
+    {
+      "github_username": "Teo-ShaoWei"
+    },
+    {
+      "github_username": "cmcaine"
+    }
+  ],
```
This is another deprecated exercise, but it appears to be correct: https://github.com/exercism/julia/commits/v2-final-version/exercises/transpose

SaschaMann commented on exercises/practice/transpose/.meta/config.json:
```diff
@@ -1,10 +1,32 @@
+      "github_username": "SaschaMann"
+    },
+    {
+      "github_username": "kytrinyx"
```
> Bulk update PRs made by ErikSchierboom or kytrinx to update the track

I believe this bulk PR may be counted by accident: https://github.com/exercism/julia/pull/75

Katrina is listed as contributor here but her only change to `transpose` was the bulk PR above, see https://github.com/exercism/julia/commits/v2-final-version/exercises/transpose

SaschaMann reviewed (commented):
I only spot-checked some exercises and found some issues. Most of it seems fine, though.

SaschaMann committed:
Fix invalid forked_from values (#356)

SaschaMann committed:
Add concept exercise blurbs (#355)

* Add concept exercise blurbs

* Update config.json

SaschaMann committed:
Add concept exercise: documented-lasagna / docstrings (#350)

Co-authored-by: Colin Caine <cmcaine@gmail.com>
Co-authored-by: Angelika Tyborska <angelikatyborska@fastmail.com>

ErikSchierboom committed:
Add authors and contributors to Practice Exercises

For each Practice Exercise, we've looked at the commit history of its files to see which commits touched that Practice Exercise.

Each commit that we could associate with a Practice Exercise was used to determine authorship/contributorship.

### Authors

1. Find the Git commit author of the oldest commit linked to that Practice Exercise
2. Find the GitHub username for the Git commit author of that commit
3. Add the GitHub username of the Git commit author to the `authors` key in the `.meta/config.json` file

### Contributors

1. Find the Git commit authors and any co-authors of all but the oldest commit linked to that Practice Exercise. If there is only one commit, there won't be any contributors.
1. Exclude the Git commit author and any co-authors of the oldest commit from the list of Git commit authors (an author is _not_ also a contributor)
2. Find the GitHub usernames for the Git commit authors and any co-authors of those commits
3. Add the GitHub usernames of the Git commit authors and any co-authors to the `contributor` key in the `.meta/config.json` file

We used the GitHub GraphQL API to find the username of a commit author or any co-authors. In some cases though, a username cannot be found for a commit (e.g. due to the user account no longer existing), in which case the commit was skipped.

## Renames

There are a small number of Practice Exercises that might have been renamed at some point. You can ask Git to "follow" a file over its renames using `git log --follow <file>`, which will also return commits made before renames. Unfortunately, Git does not store renames, it just stores the contents of the renamed files and tries to guess if a file was renamed by looking at its contents. This _can_ (and will) lead to false positives, where Git will think a file has been renamed whereas it hasn't. As we don't want to have incorrect authors/contributors for exercises, we're ignoring renames. The only exception to this are known exercise renames:

- `bracket-push` was renamed to `matching-brackets`
- `retree` was renamed to `satellite`
- `resistor-colors` was renamed to `resistor-color-duo`
- `kindergarden-garden` was renamed to `kindergarten-garden`

## Exclusions

There are some commits that we skipped over, and which thus didn't influence the authors/contributors list:

- Commits authored by `dependabot[bot]`, `dependabot-preview[bot]` or `github-actions[bot]`
- Bulk update PRs made by `ErikSchierboom` or `kytrinx` to update the track

SaschaMann committed:
Restore formatting

SaschaMann committed:
Manually fix authors

A few noteworthy changes:

* #162 has manually been excluded as it was a mass-PR done by changing a global template

* PRs to several exercises that changed the exercise itself were all included

* All reviews were included as contribution; reviewers where added where missing; however due to GitHub's UI hiding reviewers in the history, I may have missed reviews done on PRs that weren't on the initial PR of an exercise

* 6c8a1db caused weirdness due to having two parents from merging the default branch into the PR branch. This has manually been resolved as well

If you feel like you've been removed or added unjustly, let me know and I'll re-add you in a follow-up PR. Sorry if that happened!
